# 📊 Assignment 1: Python Basics & Data Exploration with Pandas

**Objective:** Learn Python basics and perform basic data exploration and cleaning using Pandas.

**Dataset:** Myntra E-Commerce Products (`Combined_dataset.csv`)

**Steps Covered:**
1. Load a CSV dataset into a Pandas DataFrame
2. Explore data (head/tail, shape, columns, data types)
3. Handle missing values (identify, fill/drop)
4. Perform basic operations (filter rows, select columns)
5. Remove duplicates
6. Create a derived column (`total_amount = price × quantity`)
7. Save the cleaned dataset as a new CSV file

---

## Import Required Libraries

In [4]:
# Import the essential library for data manipulation
import pandas as pd
import numpy as np

# Display settings for better readability
pd.set_option('display.max_columns', None)  # Show all columns
pd.set_option('display.max_colwidth', 50)    # Limit column width for readability
pd.set_option('display.width', 200)          # Wider display

print("✅ Libraries imported successfully!")
print(f"Pandas version: {pd.__version__}")
print(f"NumPy version:  {np.__version__}")

✅ Libraries imported successfully!
Pandas version: 2.3.2
NumPy version:  2.2.3


---
##  Step 1: Load a CSV Dataset into a Pandas DataFrame

We use `pd.read_csv()` to load the **Myntra E-Commerce Combined Dataset** which contains 1,000 product listings across various categories.

In [5]:
# Load the CSV file into a DataFrame
df = pd.read_csv("/Users/sandeshkadam/Desktop/CEI/archive/Combined_dataset.csv")

print(f"✅ Dataset loaded successfully!")
print(f"Shape: {df.shape[0]} rows × {df.shape[1]} columns")

✅ Dataset loaded successfully!
Shape: 1000 rows × 24 columns


---
## Step 2: Explore the Data

Data exploration is the first step in understanding the structure, content, and quality of a dataset. We'll examine:
- **head() / tail()** — First and last few rows
- **shape** — Number of rows and columns
- **columns** — Column names
- **dtypes** — Data types of each column
- **info()** — Summary of the DataFrame
- **describe()** — Statistical summary of numerical columns

In [6]:
# Display the first 5 rows to get a quick look at the data
df.head()

,product_id,title,product_description,rating,ratings_count,initial_price,discount,final_price,currency,images,delivery_options,product_details,breadcrumbs,product_specifications,amount_of_stars,what_customers_said,seller_name,sizes,videos,seller_information,variations,best_offer,more_offers,category
0,8376765,Lino Perros,Women Navy Blue Solid Backpack,3.8,15,3995,58.0,"""₹3,995.00""",INR,http://assets.myntassets.com/assets/images/837...,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Navy Blue solid backpackNon-Pa...","[{""name"":""Accessories"",""url"":""https://www.mynt...","[{""specification_name"":""Add-Ons"",""specificatio...","{""1_star"":2,""2_stars"":0,""3_stars"":3,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""rw-8376765_cae700""]",NaN,[{}],{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
1,9136281,Tommy Hilfiger,Unisex Navy Blue Striped Backpack,4.5,67,2899,35.0,"""₹2,899.00""",INR,http://assets.myntassets.com/assets/images/913...,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Navy Blue backpackNon-Padded h...","[{""name"":""Accessories"",""url"":""https://www.mynt...","[{""specification_name"":""Add-Ons"",""specificatio...","{""1_star"":3,""2_stars"":4,""3_stars"":2,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""rw-9136281_cae700""]",NaN,"[{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
2,17633752,Lavie,Aries Women Pink Mini Backpack,4.4,226,2999,65.0,"""₹2,999.00""",INR,http://assets.myntassets.com/assets/images/176...,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Pink solid backpacks Non-padd...","[{""name"":""Accessories"",""url"":""https://www.mynt...","[{""specification_name"":""Add-Ons"",""specificatio...","{""1_star"":9,""2_stars"":5,""3_stars"":10,""4_stars""...",NaN,NaN,"[{""size"":""S""}]","[""https://videos.myntassets.com/assets/videos/...",NaN,"[{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
3,1376949,F Gear,Unisex Navy & Grey Printed Burner Backpack,4.4,1052,1675,52.0,"""₹1,675.00""",INR,http://assets.myntassets.com/assets/images/137...,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Navy and grey printed backpack...","[{""name"":""Accessories"",""url"":""https://www.mynt...","[{""specification_name"":""Compartment Closure"",""...","{""1_star"":33,""2_stars"":21,""3_stars"":69,""4_star...",NaN,NaN,"[{""size"":""Onesize""}]","[""https://videos.myntassets.com/assets/videos/...",NaN,NaN,{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
4,13939916,MYTRIDENT,Men Blue Solid Bath Robe,4.7,12,2899,17.0,"""₹2,899.00""",INR,http://assets.myntassets.com/assets/images/pro...,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Blue solid bath robe, has a sh...","[{""name"":""Men"",""url"":""https://www.myntra.com/m...","[{""specification_name"":""Body or Garment Size"",...","{""1_star"":0,""2_stars"":0,""3_stars"":0,""4_stars"":...",NaN,NaN,"[{""size"":""M"",""value"":""45.0in"",""value_name"":""Fr...",NaN,NaN,"[{},{},{},{},{},{},{},{},{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on IndusI...",bath-robe


In [7]:
# Display the last 5 rows
df.tail()

,product_id,title,product_description,rating,ratings_count,initial_price,discount,final_price,currency,images,delivery_options,product_details,breadcrumbs,product_specifications,amount_of_stars,what_customers_said,seller_name,sizes,videos,seller_information,variations,best_offer,more_offers,category
995,13698258,KREDO,Men FORTE Silver Automatic Skeleton Analogue W...,3.5,11,11890,50.0,"""₹11,890.00""",INR,http://assets.myntassets.com/assets/images/136...,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Features:Nothing about the For...","[{""name"":""Accessories"",""url"":""https://www.mynt...","[{""specification_name"":""Add-Ons"",""specificatio...","{""1_star"":3,""2_stars"":0,""3_stars"":2,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""pi-13698258""]",NaN,NaN,{},"[{""offer_name"":""10% Instant Discount on Citi C...",watches
996,16896006,GUESS,Women White Patterned Dial & White Straps Anal...,5.0,22,8900,45.0,"""₹8,900.00""",INR,http://assets.myntassets.com/assets/images/168...,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Pack Contains: single watchDis...","[{""name"":""Accessories"",""url"":""https://www.mynt...","[{""specification_name"":""Add-Ons"",""specificatio...","{""1_star"":0,""2_stars"":0,""3_stars"":0,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""rw-16896006_cae700"",""https://videos.myntasse...",NaN,NaN,{},"[{""offer_name"":""10% Instant Discount on Citi C...",watches
997,1708108,GIORDANO,Women Rose Gold-Toned Mirrored Analogue Watch ...,2.8,5,7950,50.0,"""₹7,950.00""",INR,http://assets.myntassets.com/assets/images/170...,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Case style: Analogue watch, ha...","[{""name"":""Accessories"",""url"":""https://www.mynt...","[{""specification_name"":""Dial Material"",""specif...","{""1_star"":2,""2_stars"":1,""3_stars"":0,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]",NaN,NaN,"[{},{},{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",watches
998,19280114,Fastrack,Men Black Brass Dial & Black Stainless Steel S...,4.3,17,3295,20.0,"""₹2,636.00""",INR,http://assets.myntassets.com/assets/images/192...,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Pack Contains: single watch D...","[{""name"":""Accessories"",""url"":""https://www.mynt...","[{""specification_name"":""Add-Ons"",""specificatio...","{""1_star"":1,""2_stars"":2,""3_stars"":0,""4_stars"":...",NaN,Tru███m R███il,"[{""size"":""Onesize""}]",NaN,Sho███o. ███-10█████████enu███████████████████...,NaN,"{""applicable_on"":""Orders above Rs. 1499 (only ...","[{""offer_name"":""10% Instant Discount on IndusI...",watches
999,18601682,BOLDFIT,Purple Textured Yoga Mat,3.8,161,699,31.0,"""₹699.00""",INR,http://assets.myntassets.com/assets/images/186...,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Set content: One yoga matColou...","[{""name"":""Unisex"",""url"":""https://www.myntra.co...","[{""specification_name"":""Mat Thickness"",""specif...","{""1_star"":24,""2_stars"":8,""3_stars"":22,""4_stars...",NaN,NaN,"[{""size"":""Medium""}]",NaN,NaN,"[{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",yoga-mats


In [8]:
# Get the number of rows and columns
print(f"Number of Rows:    {df.shape[0]}")
print(f"Number of Columns: {df.shape[1]}")
print(f"Total Data Points: {df.shape[0] * df.shape[1]}")

Number of Rows:    1000
Number of Columns: 24
Total Data Points: 24000


In [9]:
# List all column names
print("Columns in the dataset:")
print("=" * 40)
for i, col in enumerate(df.columns, 1):
    print(f"  {i:2d}. {col}")

Columns in the dataset:
   1. product_id
   2. title
   3. product_description
   4. rating
   5. ratings_count
   6. initial_price
   7. discount
   8. final_price
   9. currency
  10. images
  11. delivery_options
  12. product_details
  13. breadcrumbs
  14. product_specifications
  15. amount_of_stars
  16. what_customers_said
  17. seller_name
  18. sizes
  19. videos
  20. seller_information
  21. variations
  22. best_offer
  23. more_offers
  24. category


In [10]:
# Check data types for each column
print("Data types of each column:")
print("=" * 40)
print(df.dtypes)

Data types of each column:
product_id                  int64
title                      object
product_description        object
rating                    float64
ratings_count               int64
initial_price               int64
discount                  float64
final_price                object
currency                   object
images                     object
delivery_options           object
product_details            object
breadcrumbs                object
product_specifications     object
amount_of_stars            object
what_customers_said        object
seller_name                object
sizes                      object
videos                     object
seller_information         object
variations                 object
best_offer                 object
more_offers                object
category                   object
dtype: object


In [11]:
# Get a concise summary of the DataFrame
# This shows column names, non-null counts, and data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   product_id              1000 non-null   int64  
 1   title                   1000 non-null   object 
 2   product_description     1000 non-null   object 
 3   rating                  1000 non-null   float64
 4   ratings_count           1000 non-null   int64  
 5   initial_price           1000 non-null   int64  
 6   discount                879 non-null    float64
 7   final_price             1000 non-null   object 
 8   currency                1000 non-null   object 
 9   images                  1000 non-null   object 
 10  delivery_options        1000 non-null   object 
 11  product_details         1000 non-null   object 
 12  breadcrumbs             1000 non-null   object 
 13  product_specifications  1000 non-null   object 
 14  amount_of_stars         1000 non-null   o

In [12]:
# Get statistical summary of numerical columns
# Shows count, mean, std, min, 25%, 50%, 75%, max
df.describe()

,product_id,rating,ratings_count,initial_price,discount
count,1.000000e+03,1000.0000,1000.000000,1000.00000,879.000000
mean,1.713437e+07,3.6215,76.691000,2723.24100,53.503982
std,3.813766e+06,1.3744,241.114263,2408.69736,18.029201
min,5.868460e+05,0.0000,0.000000,249.00000,1.000000
25%,1.511501e+07,3.7000,7.000000,1399.00000,44.000000
50%,1.820890e+07,4.1000,17.000000,1999.00000,56.000000
75%,1.902737e+07,4.3000,58.000000,3299.00000,66.000000
max,2.274707e+07,5.0000,4441.000000,22199.00000,88.000000


In [13]:
# Also describe non-numerical (object/string) columns
df.describe(include='object')

,title,product_description,final_price,currency,images,delivery_options,product_details,breadcrumbs,product_specifications,amount_of_stars,what_customers_said,seller_name,sizes,videos,seller_information,variations,best_offer,more_offers,category
count,1000,1000,1000,1000,1000,1000,1000,1000,1000,1000,427,699,1000,219,699,438,1000,1000,1000
unique,454,989,466,1,1000,16,998,658,967,808,302,301,672,219,298,22,468,159,97
top,Roadster,Women Pink Solid Top,"""₹999.00""",INR,http://assets.myntassets.com/assets/images/837...,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""A pair of navy blue running sp...","[{""name"":""Clothing"",""url"":""https://www.myntra....","[{""specification_name"":""Blouse"",""specification...","{""1_star"":0,""2_stars"":0,""3_stars"":0,""4_stars"":...","[{""percentage"":""100"",""value_name"":""Fit, Just R...",Sup███om ███,"[{""size"":""Onesize""}]","[""rw-8376765_cae700""]",Plo███o- ███ 1s█████████Sui███████████████████...,[{}],{},"[{""offer_name"":""10% Instant Discount on IndusI...",tops
freq,30,3,25,1000,1,650,2,16,4,114,27,40,135,1,40,126,306,265,122


In [14]:
# Check unique value counts for categorical-like columns
print(f"Unique brands (title):     {df['title'].nunique()}")
print(f"Unique categories:         {df['category'].nunique()}")
print(f"Unique currencies:         {df['currency'].nunique()}")

print("\nTop 10 Product Categories:")
print(df['category'].value_counts().head(10))

Unique brands (title):     454
Unique categories:         97
Unique currencies:         1

Top 10 Product Categories:
category
tops            122
dresses         100
shirts           97
jeans            57
sports-shoes     51
tshirts          39
earrings         34
sweaters         34
jackets          29
casual-shoes     23
Name: count, dtype: int64


---
## 🩹 Step 3: Handle Missing Values

Missing values can cause errors in analysis. We need to:
1. **Identify** where missing values exist
2. **Decide** whether to fill or drop them

### 3.1 Identify Missing Values

In [15]:
# Count missing values per column
missing_counts = df.isnull().sum()
missing_percent = (df.isnull().sum() / len(df) * 100).round(2)

# Create a summary table of missing values
missing_df = pd.DataFrame({
    'Missing Count': missing_counts,
    'Missing %': missing_percent
})

# Display only columns that have missing values
print("📋 Missing Values Summary:")
print("=" * 50)
missing_with_values = missing_df[missing_df['Missing Count'] > 0].sort_values('Missing Count', ascending=False)
print(missing_with_values)
print(f"\nTotal missing values in the dataset: {df.isnull().sum().sum()}")

📋 Missing Values Summary:
                     Missing Count  Missing %
videos                         781       78.1
what_customers_said            573       57.3
variations                     562       56.2
seller_name                    301       30.1
seller_information             301       30.1
discount                       121       12.1

Total missing values in the dataset: 2639


In [16]:
# Visualize missing values with a simple bar chart
if not missing_with_values.empty:
    print("\n📊 Missing Values Bar Chart:")
    print("-" * 50)
    for col, row in missing_with_values.iterrows():
        bar_length = int(row['Missing %'] / 2)
        bar = '█' * bar_length + '░' * (50 - bar_length)
        print(f"  {col:<25s} |{bar}| {row['Missing %']:5.1f}% ({int(row['Missing Count'])} nulls)")


📊 Missing Values Bar Chart:
--------------------------------------------------
  videos                    |███████████████████████████████████████░░░░░░░░░░░|  78.1% (781 nulls)
  what_customers_said       |████████████████████████████░░░░░░░░░░░░░░░░░░░░░░|  57.3% (573 nulls)
  variations                |████████████████████████████░░░░░░░░░░░░░░░░░░░░░░|  56.2% (562 nulls)
  seller_name               |███████████████░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░|  30.1% (301 nulls)
  seller_information        |███████████████░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░|  30.1% (301 nulls)
  discount                  |██████░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░|  12.1% (121 nulls)


### 3.2 Handle Missing Values — Fill & Drop Strategy

**Strategy:**
- **`discount`** (numerical, 121 missing) → Fill with **0** (no discount applied)
- **`seller_name`** (text, 301 missing) → Fill with **"Unknown"**
- **`what_customers_said`** (text, 573 missing) → Fill with **"No reviews"**
- **`videos`** (text, 781 missing) → Drop this column entirely (>78% missing, not useful for analysis)
- **`seller_information`** (text, 301 missing) → Fill with **"Not Available"**
- **`variations`** (text, 562 missing) → Fill with **"No variations"**

In [17]:
# --- FILL missing values ---

# Fill 'discount' with 0 (no discount)
df['discount'] = df['discount'].fillna(0)
print(f"'discount' — filled NaN with 0")

# Fill 'seller_name' with 'Unknown'
df['seller_name'] = df['seller_name'].fillna('Unknown')
print(f"'seller_name' — filled NaN with 'Unknown'")

# Fill 'what_customers_said' with 'No reviews'
df['what_customers_said'] = df['what_customers_said'].fillna('No reviews')
print(f"'what_customers_said' — filled NaN with 'No reviews'")

# Fill 'seller_information' with 'Not Available'
df['seller_information'] = df['seller_information'].fillna('Not Available')
print(f"'seller_information' — filled NaN with 'Not Available'")

# Fill 'variations' with 'No variations'
df['variations'] = df['variations'].fillna('No variations')
print(f"'variations' — filled NaN with 'No variations'")

'discount' — filled NaN with 0
'seller_name' — filled NaN with 'Unknown'
'what_customers_said' — filled NaN with 'No reviews'
'seller_information' — filled NaN with 'Not Available'
'variations' — filled NaN with 'No variations'


In [15]:
# --- DROP columns with too many missing values ---

# Drop 'videos' column (>78% missing, not useful for analysis)
df = df.drop(columns=['videos'])
print(f"✅ 'videos' column dropped (>78% values were missing)")

✅ 'videos' column dropped (>78% values were missing)


In [20]:
# Verify: No more missing values
remaining_missing = df.isnull().sum().sum()
print(f"\nRemaining missing values: {remaining_missing}")
if remaining_missing == 0:
    print("All missing values have been handled successfully!")
else:
    print("Some missing values still remain:")
    print(df.isnull().sum()[df.isnull().sum() > 0])


Remaining missing values: 781
Some missing values still remain:
videos    781
dtype: int64


---
## ⚙️ Step 4: Perform Basic Operations

Demonstrate key Pandas operations:
- **Select specific columns**
- **Filter rows** based on conditions

### 4.1 Select Specific Columns

In [17]:
# Select only the key columns for analysis
key_columns = ['product_id', 'title', 'product_description', 'rating', 
               'ratings_count', 'initial_price', 'discount', 'category']

df_key = df[key_columns]

print(f"Selected {len(key_columns)} key columns:")
print(key_columns)
print()
df_key.head(10)

Selected 8 key columns:
['product_id', 'title', 'product_description', 'rating', 'ratings_count', 'initial_price', 'discount', 'category']



,product_id,title,product_description,rating,ratings_count,initial_price,discount,category
0,8376765,Lino Perros,Women Navy Blue Solid Backpack,3.8,15,3995,58.0,backpacks
1,9136281,Tommy Hilfiger,Unisex Navy Blue Striped Backpack,4.5,67,2899,35.0,backpacks
2,17633752,Lavie,Aries Women Pink Mini Backpack,4.4,226,2999,65.0,backpacks
3,1376949,F Gear,Unisex Navy & Grey Printed Burner Backpack,4.4,1052,1675,52.0,backpacks
4,13939916,MYTRIDENT,Men Blue Solid Bath Robe,4.7,12,2899,17.0,bath-robe
5,17198778,H&M,Beige & Tan Solid Stoneware Soap Dispenser,4.5,42,1399,0.0,bathroom-accessories
6,19851824,AVI Living,Grey Solid 500 GSM Pure Cotton Bath Towel,4.0,5,1399,65.0,bath-towels
7,18602872,My Room,Yellow & Green Floral 140 TC King Bedsheet wit...,4.7,10,2999,54.0,bedsheets
8,18602850,My Room,Ethnic Print 144 TC Super King Bedsheet with 2...,4.5,22,2999,54.0,bedsheets
9,8961147,Story@home,Pink & White Floral 160 TC Cotton 1 King Bedsh...,3.7,7,2199,53.0,bedsheets


### 4.2 Filter Rows — Products with Rating ≥ 4.0

In [18]:
# Filter products with rating >= 4.0
high_rated = df[df['rating'] >= 4.0]

print(f"Products with rating ≥ 4.0: {len(high_rated)} out of {len(df)} ({len(high_rated)/len(df)*100:.1f}%)")
print()
high_rated[['product_id', 'title', 'product_description', 'rating', 'initial_price', 'category']].head(10)

Products with rating ≥ 4.0: 615 out of 1000 (61.5%)



,product_id,title,product_description,rating,initial_price,category
1,9136281,Tommy Hilfiger,Unisex Navy Blue Striped Backpack,4.5,2899,backpacks
2,17633752,Lavie,Aries Women Pink Mini Backpack,4.4,2999,backpacks
3,1376949,F Gear,Unisex Navy & Grey Printed Burner Backpack,4.4,1675,backpacks
4,13939916,MYTRIDENT,Men Blue Solid Bath Robe,4.7,2899,bath-robe
5,17198778,H&M,Beige & Tan Solid Stoneware Soap Dispenser,4.5,1399,bathroom-accessories
6,19851824,AVI Living,Grey Solid 500 GSM Pure Cotton Bath Towel,4.0,1399,bath-towels
7,18602872,My Room,Yellow & Green Floral 140 TC King Bedsheet wit...,4.7,2999,bedsheets
8,18602850,My Room,Ethnic Print 144 TC Super King Bedsheet with 2...,4.5,2999,bedsheets
10,8430275,MYTRIDENT,Sea Green Floral Flat 120 TC Cotton 1 Queen Be...,4.4,2199,bedsheets
15,19788630,Home Ecstasy,140tc Pure Cotton printed Teal Blue Single Bed...,4.5,1199,bedsheets


### 4.3 Filter Rows — Products with Discount > 50% and Price < ₹2000

In [19]:
# Filter products with high discount AND affordable price
budget_deals = df[(df['discount'] > 50) & (df['initial_price'] < 2000)]

print(f"Budget deals (discount > 50% & price < ₹2000): {len(budget_deals)} products")
print()
budget_deals[['product_id', 'title', 'product_description', 'rating', 
              'initial_price', 'discount', 'category']].head(10)

Budget deals (discount > 50% & price < ₹2000): 257 products



,product_id,title,product_description,rating,initial_price,discount,category
3,1376949,F Gear,Unisex Navy & Grey Printed Burner Backpack,4.4,1675,52.0,backpacks
6,19851824,AVI Living,Grey Solid 500 GSM Pure Cotton Bath Towel,4.0,1399,65.0,bath-towels
12,18327434,BELLA CASA,Unisex Blue Bedsheets,0.0,1499,73.0,bedsheets
18,22409156,Kastner,Women Textured Tang Slim Belt,3.8,1399,84.0,belts
23,17372928,The Souled Store,Men Blue & White Abstract Printed Boxers,4.5,799,54.0,boxers
39,21573578,Jewels Galaxy,Women Brass Gold-Plated Bangle-Style Bracelet,4.2,1329,77.0,bracelet
40,18100576,MINUTIAE,Women Gold-Toned & White Brass Crystals Handcr...,3.2,1499,70.0,bracelet
47,19206416,ABELINO,Women Set of 3 Assorted Cotton Boy Shorts Brie...,3.6,1899,80.0,briefs
52,22436508,Bruchi CLUB,Men Low-Rise Anti-Bacterial Basic Briefs,5.0,888,65.0,briefs
56,13669894,Vastraa Fusion,Unisex Maroon Solid Woollen Afghani Pakol Cap,3.5,1699,70.0,caps


### 4.4 Filter Rows — Products in a Specific Category

In [20]:
# Filter products in the 'backpacks' category
backpacks = df[df['category'] == 'backpacks']

print(f"Products in 'backpacks' category: {len(backpacks)}")
print(f"Average rating: {backpacks['rating'].mean():.2f}")
print(f"Average price:  ₹{backpacks['initial_price'].mean():.2f}")
print()
backpacks[['product_id', 'title', 'product_description', 'rating', 'initial_price']].head()

Products in 'backpacks' category: 4
Average rating: 4.28
Average price:  ₹2892.00



,product_id,title,product_description,rating,initial_price
0,8376765,Lino Perros,Women Navy Blue Solid Backpack,3.8,3995
1,9136281,Tommy Hilfiger,Unisex Navy Blue Striped Backpack,4.5,2899
2,17633752,Lavie,Aries Women Pink Mini Backpack,4.4,2999
3,1376949,F Gear,Unisex Navy & Grey Printed Burner Backpack,4.4,1675


---
## Step 5: Remove Duplicates



In [21]:
# Check for duplicate rows
total_duplicates = df.duplicated().sum()
print(f"Total duplicate rows found: {total_duplicates}")

# Also check for duplicate product_ids specifically
duplicate_ids = df['product_id'].duplicated().sum()
print(f"Duplicate product_id values:  {duplicate_ids}")

Total duplicate rows found: 0
Duplicate product_id values:  0


In [23]:
# Reset index after removing duplicates
df = df.reset_index(drop=True)
print(f"✅ Index reset. DataFrame shape: {df.shape}")

✅ Index reset. DataFrame shape: (1000, 23)


---
## 🧮 Step 6: Create a Derived Column

**Task:** Create `total_amount = price × quantity`

Since our dataset doesn't have a direct `quantity` column (each row is a single product listing), we adapt this step as follows:

1. **`discounted_price`** = `initial_price × (1 - discount/100)` — the actual price after discount
2. **Simulated `quantity`** — We simulate a `quantity` column (random values 1-5) to demonstrate the formula
3. **`total_amount`** = `discounted_price × quantity`

In [21]:
# Step 6a: Create 'discounted_price' column from initial_price and discount
df['discounted_price'] = (df['initial_price'] * (1 - df['discount'] / 100)).round(2)

print(" Created 'discounted_price' column")
print(f"   Formula: discounted_price = initial_price × (1 - discount/100)")
print()
print(df[['product_id', 'title', 'initial_price', 'discount', 'discounted_price']].head(10))

 Created 'discounted_price' column
   Formula: discounted_price = initial_price × (1 - discount/100)

   product_id           title  initial_price  discount  discounted_price
0     8376765     Lino Perros           3995      58.0           1677.90
1     9136281  Tommy Hilfiger           2899      35.0           1884.35
2    17633752           Lavie           2999      65.0           1049.65
3     1376949          F Gear           1675      52.0            804.00
4    13939916       MYTRIDENT           2899      17.0           2406.17
5    17198778             H&M           1399       0.0           1399.00
6    19851824      AVI Living           1399      65.0            489.65
7    18602872         My Room           2999      54.0           1379.54
8    18602850         My Room           2999      54.0           1379.54
9     8961147      Story@home           2199      53.0           1033.53


In [22]:
# Step 6b: Simulate a 'quantity' column (1 to 5 units)
np.random.seed(42)  # For reproducibility
df['quantity'] = np.random.randint(1, 6, size=len(df))

print(" Created 'quantity' column (simulated: random values 1–5)")
print(f"   Quantity distribution:")
print(df['quantity'].value_counts().sort_index())

 Created 'quantity' column (simulated: random values 1–5)
   Quantity distribution:
quantity
1    210
2    190
3    190
4    206
5    204
Name: count, dtype: int64


In [25]:
# Step 6c: Create 'total_amount' = discounted_price × quantity
df['total_amount'] = (df['discounted_price'] * df['quantity']).round(2)

print("✅ Created 'total_amount' column")
print(f"   Formula: total_amount = discounted_price × quantity")
print()
print(df[['product_id', 'title', 'initial_price', 'discount', 
          'discounted_price', 'quantity', 'total_amount']].head(10))

✅ Created 'total_amount' column
   Formula: total_amount = discounted_price × quantity

   product_id           title  initial_price  discount  discounted_price  quantity  total_amount
0     8376765     Lino Perros           3995      58.0           1677.90         4       6711.60
1     9136281  Tommy Hilfiger           2899      35.0           1884.35         5       9421.75
2    17633752           Lavie           2999      65.0           1049.65         3       3148.95
3     1376949          F Gear           1675      52.0            804.00         5       4020.00
4    13939916       MYTRIDENT           2899      17.0           2406.17         5      12030.85
5    17198778             H&M           1399       0.0           1399.00         2       2798.00
6    19851824      AVI Living           1399      65.0            489.65         3       1468.95
7    18602872         My Room           2999      54.0           1379.54         3       4138.62
8    18602850         My Room          

In [27]:
# Summary statistics of the new derived columns
print("Summary Statistics of Derived Columns:")
print("=" * 50)
print(df[['initial_price', 'discount', 'discounted_price', 'quantity', 'total_amount']].describe().round(2))

Summary Statistics of Derived Columns:
       initial_price  discount  discounted_price  quantity  total_amount
count        1000.00   1000.00           1000.00   1000.00       1000.00
mean         2723.24     47.03           1410.08      3.00       4113.60
std          2408.70     24.30           1618.96      1.43       4818.77
min           249.00      0.00            199.60      1.00        219.56
25%          1399.00     33.00            571.77      2.00       1362.36
50%          1999.00     53.00            874.75      3.00       2598.85
75%          3299.00     65.00           1599.09      4.00       4980.72
max         22199.00     88.00          16999.00      5.00      53996.00


---
## 💾 Step 7: Save the Cleaned Dataset as a New CSV File

Save the final cleaned and enriched DataFrame to a new CSV file.

In [28]:
# Final check before saving
print("📋 Final Dataset Summary:")
print("=" * 50)
print(f"Shape:               {df.shape[0]} rows × {df.shape[1]} columns")
print(f"Missing values:      {df.isnull().sum().sum()}")
print(f"Duplicate rows:      {df.duplicated().sum()}")
print(f"Columns:             {list(df.columns)}")
print()
df.info()

📋 Final Dataset Summary:
Shape:               1000 rows × 27 columns
Missing values:      781
Duplicate rows:      0
Columns:             ['product_id', 'title', 'product_description', 'rating', 'ratings_count', 'initial_price', 'discount', 'final_price', 'currency', 'images', 'delivery_options', 'product_details', 'breadcrumbs', 'product_specifications', 'amount_of_stars', 'what_customers_said', 'seller_name', 'sizes', 'videos', 'seller_information', 'variations', 'best_offer', 'more_offers', 'category', 'discounted_price', 'quantity', 'total_amount']

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 27 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   product_id              1000 non-null   int64  
 1   title                   1000 non-null   object 
 2   product_description     1000 non-null   object 
 3   rating                  1000 non-null   float64
 4   ratings_cou

In [29]:
# Save the cleaned dataset
output_path = "/Users/sandeshkadam/Desktop/CEI/Assignment 1/Cleaned_dataset.csv"
df.to_csv(output_path, index=False)

print(f"Cleaned dataset saved successfully!")
print(f"Location: {output_path}")
print(f"Shape: {df.shape[0]} rows × {df.shape[1]} columns")

Cleaned dataset saved successfully!
Location: /Users/sandeshkadam/Desktop/CEI/Assignment 1/Cleaned_dataset.csv
Shape: 1000 rows × 27 columns


In [30]:
# Verify by reading back the saved file
df_verify = pd.read_csv(output_path)
print("🔍 Verification — Reading back the saved CSV:")
print(f"   Shape: {df_verify.shape}")
print(f"   Missing values: {df_verify.isnull().sum().sum()}")
print()
df_verify.head()

🔍 Verification — Reading back the saved CSV:
   Shape: (1000, 27)
   Missing values: 781



,product_id,title,product_description,rating,ratings_count,initial_price,discount,final_price,currency,images,delivery_options,product_details,breadcrumbs,product_specifications,amount_of_stars,what_customers_said,seller_name,sizes,videos,seller_information,variations,best_offer,more_offers,category,discounted_price,quantity,total_amount
0,8376765,Lino Perros,Women Navy Blue Solid Backpack,3.8,15,3995,58.0,"""₹3,995.00""",INR,http://assets.myntassets.com/assets/images/837...,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Navy Blue solid backpackNon-Pa...","[{""name"":""Accessories"",""url"":""https://www.mynt...","[{""specification_name"":""Add-Ons"",""specificatio...","{""1_star"":2,""2_stars"":0,""3_stars"":3,""4_stars"":...",No reviews,Unknown,"[{""size"":""Onesize""}]","[""rw-8376765_cae700""]",Not Available,[{}],{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks,1677.90,4,6711.60
1,9136281,Tommy Hilfiger,Unisex Navy Blue Striped Backpack,4.5,67,2899,35.0,"""₹2,899.00""",INR,http://assets.myntassets.com/assets/images/913...,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Navy Blue backpackNon-Padded h...","[{""name"":""Accessories"",""url"":""https://www.mynt...","[{""specification_name"":""Add-Ons"",""specificatio...","{""1_star"":3,""2_stars"":4,""3_stars"":2,""4_stars"":...",No reviews,Unknown,"[{""size"":""Onesize""}]","[""rw-9136281_cae700""]",Not Available,"[{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks,1884.35,5,9421.75
2,17633752,Lavie,Aries Women Pink Mini Backpack,4.4,226,2999,65.0,"""₹2,999.00""",INR,http://assets.myntassets.com/assets/images/176...,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Pink solid backpacks Non-padd...","[{""name"":""Accessories"",""url"":""https://www.mynt...","[{""specification_name"":""Add-Ons"",""specificatio...","{""1_star"":9,""2_stars"":5,""3_stars"":10,""4_stars""...",No reviews,Unknown,"[{""size"":""S""}]","[""https://videos.myntassets.com/assets/videos/...",Not Available,"[{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks,1049.65,3,3148.95
3,1376949,F Gear,Unisex Navy & Grey Printed Burner Backpack,4.4,1052,1675,52.0,"""₹1,675.00""",INR,http://assets.myntassets.com/assets/images/137...,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Navy and grey printed backpack...","[{""name"":""Accessories"",""url"":""https://www.mynt...","[{""specification_name"":""Compartment Closure"",""...","{""1_star"":33,""2_stars"":21,""3_stars"":69,""4_star...",No reviews,Unknown,"[{""size"":""Onesize""}]","[""https://videos.myntassets.com/assets/videos/...",Not Available,No variations,{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks,804.00,5,4020.00
4,13939916,MYTRIDENT,Men Blue Solid Bath Robe,4.7,12,2899,17.0,"""₹2,899.00""",INR,http://assets.myntassets.com/assets/images/pro...,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Blue solid bath robe, has a sh...","[{""name"":""Men"",""url"":""https://www.myntra.com/m...","[{""specification_name"":""Body or Garment Size"",...","{""1_star"":0,""2_stars"":0,""3_stars"":0,""4_stars"":...",No reviews,Unknown,"[{""size"":""M"",""value"":""45.0in"",""value_name"":""Fr...",NaN,Not Available,"[{},{},{},{},{},{},{},{},{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on IndusI...",bath-robe,2406.17,5,12030.85


---
## 📝 Brief Summary

### What was done:

| Step | Task | Details |
|------|------|---------|
| **1** | Load CSV | Loaded `Combined_dataset.csv` with **1,000 rows × 24 columns** |
| **2** | Explore Data | Used `head()`, `tail()`, `shape`, `columns`, `dtypes`, `info()`, `describe()` |
| **3** | Handle Missing Values | Filled `discount` (0), `seller_name` ("Unknown"), `what_customers_said` ("No reviews"), `seller_information` ("Not Available"), `variations` ("No variations"); Dropped `videos` column (>78% null) |
| **4** | Basic Operations | Filtered by rating ≥ 4.0, budget deals (discount > 50% & price < ₹2000), category selection |
| **5** | Remove Duplicates | Checked & removed duplicate rows and duplicate `product_id` entries |
| **6** | Derived Column | Created `discounted_price`, simulated `quantity` (1–5), computed `total_amount = discounted_price × quantity` |
| **7** | Save Cleaned CSV | Saved to `Cleaned_dataset.csv` with **0 missing values** and **0 duplicates** |

### Key Findings:
- The dataset contains e-commerce product listings from Myntra across **multiple categories** (backpacks, shirts, jeans, watches, etc.)
- **121 products** had no discount information (filled with 0)
- The `videos` column was **>78% empty** and was dropped entirely
- Average product rating is approximately **3.6/5.0**
- Average initial price is approximately **₹2,723**
- The average discount is approximately **53%**

### Output Files:
- 📓 `assignment1.ipynb` 
- 📄 `Cleaned_dataset.csv` 